[docs] assertions refactoring (#4901)
DevExpress/testcafe: docs/articles/documentation/guides/basic-guides/assert.md

---
layout: docs
title: Assert
permalink: /documentation/guides/basic-guides/assert.html
---
# Assertions

You can use *assertions* to check if the tested page's state matches the expected state.

TestCafe provides a comprehensive set of assertions that are based on the Behavior-Driven Development style (BDD-style).

This topic consists of the following sections:

* [Assertion Structure](#assertion-structure)
* [Smart Assertion Query Mechanism](#smart-assertion-query-mechanism)

## Assertion Structure

> To construct assertions, use the [test controller](../../reference/test-api/testcontroller/README.md)'s `expect` method.

Reviewer: I think we should link `[expect](../../reference/test-api/testcontroller/expect/README.md)`.

And probably remove the test controller link in this case.

});
```

The following assertion methods are available:

* [Deep Equal](../../reference/test-api/testcontroller/expect/eql.md)
* [Not Deep Equal](../../reference/test-api/testcontroller/expect/notEql.md)
* [Ok](../../reference/test-api/testcontroller/expect/ok.md)
* [Not Ok](../../reference/test-api/testcontroller/expect/notOk.md)
* [Contains](../../reference/test-api/testcontroller/expect/contains.md)
* [Not Contains](../../reference/test-api/testcontroller/expect/notContains.md)
* [Type of](../../reference/test-api/testcontroller/expect/typeOf.md)
* [Not Type of](../../reference/test-api/testcontroller/expect/notTypeOf.md)
* [Greater than](../../reference/test-api/testcontroller/expect/gt.md)
* [Greater than or Equal to](../../reference/test-api/testcontroller/expect/gte.md)
* [Less than](../../reference/test-api/testcontroller/expect/lt.md)
* [Less than or Equal to](../../reference/test-api/testcontroller/expect/lte.md)
* [Within](../../reference/test-api/testcontroller/expect/within.md)
* [Not Within](../../reference/test-api/testcontroller/expect/notWithin.md)
* [Match](../../reference/test-api/testcontroller/expect/match.md)

> * [Not Match](../../reference/test-api/testcontroller/expect/notMatch.md)

Reviewer: Please rename all files using lower-case (except for `README.md`), update all links in the markdown and permalinks in the front matter.

## Smart Assertion Query Mechanism

In synchronous functional testing, you can execute assertions immediately after test actions.

![Synchronous Functional Testing](../../../images/assertions/synchronous-testing.png)

> Functional tests are asynchronous on the web. This means that we cannot get the expected changes immediately after an end user's actions.

Reviewer 2: ```suggestion
Functional tests are asynchronous on the web. This means that we cannot get the expected changes immediately after a user's actions.
```

> For example, it can take time for the tested page to send a request to the server for the required data, or an end user's action launches an animation after which the page reaches its final state.

Reviewer 2: ```suggestion
For example, it can take time for the tested page to send a request to the server for the required data or a user's action to invoke an animation after which the page reaches its final state.
```

All these intervals cannot be pre-calculated because they depend on various factors: computer performance,
network connection speed, etc. In this case, if we perform assertions immediately after the test action, we can get an inconclusive result.

![Asynchronous Functional Testing](../../../images/assertions/asynchronous-testing.png)

An additional timeout is usually added when performing asynchronous functional tests.

![Asynchronous Functional Testing with Extra Waiting](../../../images/assertions/extra-waiting.png)

> To stabilize asynchronous functional tests, you need to add a timeout that enables the required changes to be applied successfully.

Reviewer 2: ```suggestion
To stabilize asynchronous functional tests, you need to add a timeout that allows changes to be applied.
```

if an assertion did not pass, the test does not fail immediately. The assertion retries to pass multiple times, and
each time it requests the `actual` property value. The test fails if the assertion could not complete successfully
within a timeout:

![TestCafe Smart Assertion Query Mechanism](../../../images/assertions/query-mechanism.png)

**Example:**

The following web page is an example:

```html
<div id="btn"></div>
<script>
var btn = document.getElementById('btn');

btn.addEventListener('click', function() {
    window.setTimeout(function() {
        btn.innerText = 'Loading...';
    }, 100);
});
</script>
```

Test code for this page can be as follows.

```js
test('Button click', async t => {
    const btn = Selector('#btn');

    await t
        .click(btn)
        // A regular assertion fails immediately, but TestCafe retries to run DOM state
        // assertions many times within the timeout until this assertion passes successfully.
        // The default timeout is 3000 ms.
        .expect(btn.textContent).contains('Loading...');
});
```

The approach described above allows you to create stable tests with fast runtime and reduces the risk of errors.

> You can specify the assertion query timeout in test code by using the [options.timeout](#options) option.

Reviewer: Please clean the old content of `by using` and other unnecessary -ings.

To set the timeout when you launch tests, pass the timeout value to the [runner.run](../../using-testcafe/programming-interface/runner.md#run)

> method if you use an API or specify the [assertion-timeout](../../using-testcafe/command-line-interface.md#--assertion-timeout-ms) option

Reviewer 2: ```suggestion
method if you use API or specify the [assertion-timeout](../../using-testcafe/command-line-interface.md#--assertion-timeout-ms) option
```

if you run TestCafe from the command line.

## Options

* options.timeout

> {% include assertions/timeout.md %}

Reviewer: I guess this is going to result in:

```
## Options

* options.timeout

## options.timeout

...

```

I guess we should **remove headers from the templates** and use them where we paste:

```
## Options

### options.timeout

 {% include assertions/timeout.md %}
```
